<a href="https://colab.research.google.com/github/Ellana555/Skillfactory/blob/main/Pandas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd

In [2]:
pd.__version__

'1.3.5'

In [3]:
print(pd.__name__)

pandas


# Series

In [4]:
countries = pd.Series(
    data = ['Англия', 'Канада', 'США', 'Россия', 'Украина', 'Беларусь', 'Казахстан'],
    index = ['UK', 'CA', 'US', 'RU', 'UA', 'BY', 'KZ'],
    name = 'countries'
)
display(countries)

UK       Англия
CA       Канада
US          США
RU       Россия
UA      Украина
BY     Беларусь
KZ    Казахстан
Name: countries, dtype: object

In [5]:
countries = pd.Series({
    'UK': 'Англия',
    'CA': 'Канада',
    'US' : 'США',
    'RU': 'Россия',
    'UA': 'Украина',
    'BY': 'Беларусь',
    'KZ': 'Казахстан'},
    name = 'countries'
)
display(countries)

UK       Англия
CA       Канада
US          США
RU       Россия
UA      Украина
BY     Беларусь
KZ    Казахстан
Name: countries, dtype: object

In [6]:
print(countries.loc['US'])
# США

США


In [7]:
print(countries.loc[['US', 'RU', 'UK']])

US       США
RU    Россия
UK    Англия
Name: countries, dtype: object


In [8]:
print(countries.iloc[6])
# Казахстан

Казахстан


In [9]:
print(countries.iloc[1:4])

CA    Канада
US       США
RU    Россия
Name: countries, dtype: object


In [10]:
names=['chlorhexidine', 'cyntomycin', 'afobazol']
counts=[15, 18, 7]

In [11]:
#Напишите функцию create_medications(names, counts), 
#создающую Series medications, индексами которого 
#являются названия лекарств names, а значениями — их количество в партии counts.

#Также напишите функцию get_percent(medications, name), 
#которая возвращает долю товара с именем name от общего 
#количества товаров в партии в процентах.

import pandas as pd
import numpy as np

def create_medications(names, counts):
    medications = pd.Series(
        data = counts,
        index = [names],
        name = 'medications')
    global sum_med
    sum_med = np.sum(counts)
    return medications

def get_percent(medications, name):
    count_name = medications.loc[name]
    procent = count_name / (sum_med / 100)
    return procent


if __name__ == '__main__':
    names=['chlorhexidine', 'cyntomycin', 'afobazol']
    counts=[15, 18, 7]
    medications = create_medications(names, counts)
    print(get_percent(medications, "chlorhexidine")) #37.5

names=['chlorhexidine', 'cyntomycin', 'afobazol']
counts=[15, 18, 7]

chlorhexidine    37.5
Name: medications, dtype: float64


# DataFrime

In [12]:
countries_df = pd.DataFrame({
    'country': ['Англия', 'Канада', 'США', 'Россия', 'Украина', 'Беларусь', 'Казахстан'],
    'population': [56.29, 38.05, 322.28, 146.24, 45.5, 9.5, 17.04],
    'square': [133396, 9984670, 9826630, 17125191, 603628, 207600, 2724902]
})

In [13]:
countries_df = pd.DataFrame(
    data = [
        ['Англия', 56.29, 133396],
        ['Канада', 38.05, 9984670],
        ['США', 322.28, 9826630],
        ['Россия', 146.24, 17125191],
        ['Украина', 45.5, 603628],
        ['Беларусь', 9.5, 207600],
        ['Казахстан', 17.04, 2724902]
    ],
    columns= ['country', 'population', 'square'],
    index = ['UK', 'CA', 'US', 'RU', 'UA', 'BY', 'KZ']
)
display(countries_df)

country  population    square
UK     Англия       56.29    133396
CA     Канада       38.05   9984670
US        США      322.28   9826630
RU     Россия      146.24  17125191
UA    Украина       45.50    603628
BY   Беларусь        9.50    207600
KZ  Казахстан       17.04   2724902

In [14]:
countries_df.mean(axis=0)

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  """Entry point for launching an IPython kernel.


population    9.070000e+01
square        5.800860e+06
dtype: float64

In [15]:
countries_df.mean(axis=1)

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  """Entry point for launching an IPython kernel.


UK      66726.145
CA    4992354.025
US    4913476.140
RU    8562668.620
UA     301836.750
BY     103804.750
KZ    1362459.520
dtype: float64

In [16]:
countries_df.population

UK     56.29
CA     38.05
US    322.28
RU    146.24
UA     45.50
BY      9.50
KZ     17.04
Name: population, dtype: float64

In [17]:
countries_df['population']

UK     56.29
CA     38.05
US    322.28
RU    146.24
UA     45.50
BY      9.50
KZ     17.04
Name: population, dtype: float64

In [18]:
countries_df.loc['UK', 'square']

133396

In [19]:
countries_df.loc['RU', ['population', 'square']]

population      146.24
square        17125191
Name: RU, dtype: object

In [20]:
pd.DataFrame([[0,1], [1, 0]], columns=['А', 'B'])

А  B
0  0  1
1  1  0

In [21]:
pd.DataFrame({'А': [0, 1], 'B': [1, 0]})

А  B
0  0  1
1  1  0

In [22]:
def create_companyDF(income, expenses, years):
    df = pd.DataFrame({
        'Income':income,
        'Expenses':expenses})
    df.index = years
    return df

In [23]:
income = [478, 512, 196]
expenses = [156, 130, 270]
years = [2018, 2019, 2020]
display(create_companyDF(income, expenses, years))

Income  Expenses
2018     478       156
2019     512       130
2020     196       270

In [24]:
import pandas as pd

def create_companyDF(income, expenses, years):
    df = pd.DataFrame({
        'Income':income,
        'Expenses':expenses},
        index = years)
    return df

def get_profit(df, year):
    if year in df.index:
        profit = df.loc[year, 'Income'] - df.loc[year, 'Expenses']
    else:
        profit=None
    return profit
    #df['pribil'] = df['income'] - df['expenses']
    #profit = pd.loc('year', 'pribil')
    
    """
    А также напишите функцию get_profit(df, year), которая возвращает разницу между доходом и расходом, записанных в таблице df, за год year.
    Учтите, что если информация за запрашиваемый год не указана в вашей таблице вам необходимо вернуть None. 
    """
    #ваш код здесь
    return profit

if __name__ == '__main__':
    expenses = [156, 130, 270]
    income = [478, 512, 196]
    years = [2018, 2019, 2020]
    
    scienceyou = create_companyDF(income, expenses, years)
    print(get_profit(scienceyou, 2020)) #-74


-74


Работа с csv файлом

In [25]:
data = pd.read_csv('/content/data/melb_data.csv')
display(data)

index         Suburb           Address  Rooms Type      Price Method  \
0          0     Abbotsford      85 Turner St      2    h  1480000.0      S   
1          1     Abbotsford   25 Bloomburg St      2    h  1035000.0      S   
2          2     Abbotsford      5 Charles St      3    h  1465000.0     SP   
3          3     Abbotsford  40 Federation La      3    h   850000.0     PI   
4          4     Abbotsford       55a Park St      4    h  1600000.0     VB   
...      ...            ...               ...    ...  ...        ...    ...   
13575  13575  Wheelers Hill      12 Strada Cr      4    h  1245000.0      S   
13576  13576   Williamstown     77 Merrett Dr      3    h  1031000.0     SP   
13577  13577   Williamstown       83 Power St      3    h  1170000.0      S   
13578  13578   Williamstown      96 Verdon St      4    h  2500000.0     PI   
13579  13579     Yarraville        6 Agnes St      4    h  1285000.0     SP   

        SellerG        Date  Distance  ...  Car  Landsize  BuildingArea  \
0        Biggin   3/12/2016       2.5  ...  1.0     202.0         126.0   
1        Biggin   4/02/2016       2.5  ...  0.0     156.0          79.0   
2        Biggin   4/03/2017       2.5  ...  0.0     134.0         150.0   
3        Biggin   4/03/2017       2.5  ...  1.0      94.0         126.0   
4        Nelson   4/06/2016       2.5  ...  2.0     120.0         142.0   
...         ...         ...       ...  ...  ...       ...           ...   
13575     Barry  26/08/2017      16.7  ...  2.0     652.0         126.0   
13576  Williams  26/08/2017       6.8  ...  2.0     333.0         133.0   
13577     Raine  26/08/2017       6.8  ...  4.0     436.0         126.0   
13578   Sweeney  26/08/2017       6.8  ...  5.0     866.0         157.0   
13579   Village  26/08/2017       6.3  ...  1.0     362.0         112.0   

       YearBuilt  CouncilArea  Lattitude  Longtitude  \
0         1970.0        Yarra  -37.79960   144.99840   
1         1900.0        Yarra  -37.80790   144.99340   
2         1900.0        Yarra  -37.80930   144.99440   
3         1970.0        Yarra  -37.79690   144.99690   
4         2014.0        Yarra  -37.80720   144.99410   
...          ...          ...        ...         ...   
13575     1981.0          NaN  -37.90562   145.16761   
13576     1995.0          NaN  -37.85927   144.87904   
13577     1997.0          NaN  -37.85274   144.88738   
13578     1920.0          NaN  -37.85908   144.89299   
13579     1920.0          NaN  -37.81188   144.88449   

                       Regionname  Propertycount           Coordinates  
0           Northern Metropolitan         4019.0    -37.7996, 144.9984  
1           Northern Metropolitan         4019.0    -37.8079, 144.9934  
2           Northern Metropolitan         4019.0    -37.8093, 144.9944  
3           Northern Metropolitan         4019.0    -37.7969, 144.9969  
4           Northern Metropolitan         4019.0    -37.8072, 144.9941  
...                           ...            ...                   ...  
13575  South-Eastern Metropolitan         7392.0  -37.90562, 145.16761  
13576        Western Metropolitan         6380.0  -37.85927, 144.87904  
13577        Western Metropolitan         6380.0  -37.85274, 144.88738  
13578        Western Metropolitan         6380.0  -37.85908, 144.89299  
13579        Western Metropolitan         6543.0  -37.81188, 144.88449  

[13580 rows x 23 columns]

In [26]:
melb_data = pd.read_csv('/content/data/melb_data.csv', sep=',')

In [27]:
display(melb_data)

index         Suburb           Address  Rooms Type      Price Method  \
0          0     Abbotsford      85 Turner St      2    h  1480000.0      S   
1          1     Abbotsford   25 Bloomburg St      2    h  1035000.0      S   
2          2     Abbotsford      5 Charles St      3    h  1465000.0     SP   
3          3     Abbotsford  40 Federation La      3    h   850000.0     PI   
4          4     Abbotsford       55a Park St      4    h  1600000.0     VB   
...      ...            ...               ...    ...  ...        ...    ...   
13575  13575  Wheelers Hill      12 Strada Cr      4    h  1245000.0      S   
13576  13576   Williamstown     77 Merrett Dr      3    h  1031000.0     SP   
13577  13577   Williamstown       83 Power St      3    h  1170000.0      S   
13578  13578   Williamstown      96 Verdon St      4    h  2500000.0     PI   
13579  13579     Yarraville        6 Agnes St      4    h  1285000.0     SP   

        SellerG        Date  Distance  ...  Car  Landsize  BuildingArea  \
0        Biggin   3/12/2016       2.5  ...  1.0     202.0         126.0   
1        Biggin   4/02/2016       2.5  ...  0.0     156.0          79.0   
2        Biggin   4/03/2017       2.5  ...  0.0     134.0         150.0   
3        Biggin   4/03/2017       2.5  ...  1.0      94.0         126.0   
4        Nelson   4/06/2016       2.5  ...  2.0     120.0         142.0   
...         ...         ...       ...  ...  ...       ...           ...   
13575     Barry  26/08/2017      16.7  ...  2.0     652.0         126.0   
13576  Williams  26/08/2017       6.8  ...  2.0     333.0         133.0   
13577     Raine  26/08/2017       6.8  ...  4.0     436.0         126.0   
13578   Sweeney  26/08/2017       6.8  ...  5.0     866.0         157.0   
13579   Village  26/08/2017       6.3  ...  1.0     362.0         112.0   

       YearBuilt  CouncilArea  Lattitude  Longtitude  \
0         1970.0        Yarra  -37.79960   144.99840   
1         1900.0        Yarra  -37.80790   144.99340   
2         1900.0        Yarra  -37.80930   144.99440   
3         1970.0        Yarra  -37.79690   144.99690   
4         2014.0        Yarra  -37.80720   144.99410   
...          ...          ...        ...         ...   
13575     1981.0          NaN  -37.90562   145.16761   
13576     1995.0          NaN  -37.85927   144.87904   
13577     1997.0          NaN  -37.85274   144.88738   
13578     1920.0          NaN  -37.85908   144.89299   
13579     1920.0          NaN  -37.81188   144.88449   

                       Regionname  Propertycount           Coordinates  
0           Northern Metropolitan         4019.0    -37.7996, 144.9984  
1           Northern Metropolitan         4019.0    -37.8079, 144.9934  
2           Northern Metropolitan         4019.0    -37.8093, 144.9944  
3           Northern Metropolitan         4019.0    -37.7969, 144.9969  
4           Northern Metropolitan         4019.0    -37.8072, 144.9941  
...                           ...            ...                   ...  
13575  South-Eastern Metropolitan         7392.0  -37.90562, 145.16761  
13576        Western Metropolitan         6380.0  -37.85927, 144.87904  
13577        Western Metropolitan         6380.0  -37.85274, 144.88738  
13578        Western Metropolitan         6380.0  -37.85908, 144.89299  
13579        Western Metropolitan         6543.0  -37.81188, 144.88449  

[13580 rows x 23 columns]

In [28]:
display(melb_data.head())

index      Suburb           Address  Rooms Type      Price Method SellerG  \
0      0  Abbotsford      85 Turner St      2    h  1480000.0      S  Biggin   
1      1  Abbotsford   25 Bloomburg St      2    h  1035000.0      S  Biggin   
2      2  Abbotsford      5 Charles St      3    h  1465000.0     SP  Biggin   
3      3  Abbotsford  40 Federation La      3    h   850000.0     PI  Biggin   
4      4  Abbotsford       55a Park St      4    h  1600000.0     VB  Nelson   

        Date  Distance  ...  Car  Landsize  BuildingArea  YearBuilt  \
0  3/12/2016       2.5  ...  1.0     202.0         126.0     1970.0   
1  4/02/2016       2.5  ...  0.0     156.0          79.0     1900.0   
2  4/03/2017       2.5  ...  0.0     134.0         150.0     1900.0   
3  4/03/2017       2.5  ...  1.0      94.0         126.0     1970.0   
4  4/06/2016       2.5  ...  2.0     120.0         142.0     2014.0   

   CouncilArea  Lattitude  Longtitude             Regionname  Propertycount  \
0        Yarra   -37.7996    144.9984  Northern Metropolitan         4019.0   
1        Yarra   -37.8079    144.9934  Northern Metropolitan         4019.0   
2        Yarra   -37.8093    144.9944  Northern Metropolitan         4019.0   
3        Yarra   -37.7969    144.9969  Northern Metropolitan         4019.0   
4        Yarra   -37.8072    144.9941  Northern Metropolitan         4019.0   

          Coordinates  
0  -37.7996, 144.9984  
1  -37.8079, 144.9934  
2  -37.8093, 144.9944  
3  -37.7969, 144.9969  
4  -37.8072, 144.9941  

[5 rows x 23 columns]

In [29]:
melb_data.tail(7)

index         Suburb        Address  Rooms Type      Price Method  \
13573  13573       Werribee    5 Nuragi Ct      4    h   635000.0      S   
13574  13574    Westmeadows     9 Black St      3    h   582000.0      S   
13575  13575  Wheelers Hill   12 Strada Cr      4    h  1245000.0      S   
13576  13576   Williamstown  77 Merrett Dr      3    h  1031000.0     SP   
13577  13577   Williamstown    83 Power St      3    h  1170000.0      S   
13578  13578   Williamstown   96 Verdon St      4    h  2500000.0     PI   
13579  13579     Yarraville     6 Agnes St      4    h  1285000.0     SP   

             SellerG        Date  Distance  ...  Car  Landsize  BuildingArea  \
13573  hockingstuart  26/08/2017      14.7  ...  1.0     662.0         172.0   
13574            Red  26/08/2017      16.5  ...  2.0     256.0         126.0   
13575          Barry  26/08/2017      16.7  ...  2.0     652.0         126.0   
13576       Williams  26/08/2017       6.8  ...  2.0     333.0         133.0   
13577          Raine  26/08/2017       6.8  ...  4.0     436.0         126.0   
13578        Sweeney  26/08/2017       6.8  ...  5.0     866.0         157.0   
13579        Village  26/08/2017       6.3  ...  1.0     362.0         112.0   

       YearBuilt  CouncilArea  Lattitude  Longtitude  \
13573     1980.0          NaN  -37.89327   144.64789   
13574     1970.0          NaN  -37.67917   144.89390   
13575     1981.0          NaN  -37.90562   145.16761   
13576     1995.0          NaN  -37.85927   144.87904   
13577     1997.0          NaN  -37.85274   144.88738   
13578     1920.0          NaN  -37.85908   144.89299   
13579     1920.0          NaN  -37.81188   144.88449   

                       Regionname  Propertycount           Coordinates  
13573        Western Metropolitan        16166.0  -37.89327, 144.64789  
13574       Northern Metropolitan         2474.0   -37.67917, 144.8939  
13575  South-Eastern Metropolitan         7392.0  -37.90562, 145.16761  
13576        Western Metropolitan         6380.0  -37.85927, 144.87904  
13577        Western Metropolitan         6380.0  -37.85274, 144.88738  
13578        Western Metropolitan         6380.0  -37.85908, 144.89299  
13579        Western Metropolitan         6543.0  -37.81188, 144.88449  

[7 rows x 23 columns]

In [30]:
 melb_data.shape

(13580, 23)

In [31]:
melb_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13580 entries, 0 to 13579
Data columns (total 23 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   index          13580 non-null  int64  
 1   Suburb         13580 non-null  object 
 2   Address        13580 non-null  object 
 3   Rooms          13580 non-null  int64  
 4   Type           13580 non-null  object 
 5   Price          13580 non-null  float64
 6   Method         13580 non-null  object 
 7   SellerG        13580 non-null  object 
 8   Date           13580 non-null  object 
 9   Distance       13580 non-null  float64
 10  Postcode       13580 non-null  int64  
 11  Bedroom        13580 non-null  float64
 12  Bathroom       13580 non-null  float64
 13  Car            13580 non-null  float64
 14  Landsize       13580 non-null  float64
 15  BuildingArea   13580 non-null  float64
 16  YearBuilt      13580 non-null  float64
 17  CouncilArea    12211 non-null  object 
 18  Lattit

In [32]:
melb_data['Car'] = melb_data['Car'].astype('int64')
melb_data['Bedroom'] = melb_data['Bedroom'].astype('int64')
melb_data['Bathroom'] = melb_data['Bathroom'].astype('int64')
melb_data['Propertycount'] = melb_data['Propertycount'].astype('int64')
melb_data['YearBuilt'] = melb_data['YearBuilt'].astype('int64')
melb_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13580 entries, 0 to 13579
Data columns (total 23 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   index          13580 non-null  int64  
 1   Suburb         13580 non-null  object 
 2   Address        13580 non-null  object 
 3   Rooms          13580 non-null  int64  
 4   Type           13580 non-null  object 
 5   Price          13580 non-null  float64
 6   Method         13580 non-null  object 
 7   SellerG        13580 non-null  object 
 8   Date           13580 non-null  object 
 9   Distance       13580 non-null  float64
 10  Postcode       13580 non-null  int64  
 11  Bedroom        13580 non-null  int64  
 12  Bathroom       13580 non-null  int64  
 13  Car            13580 non-null  int64  
 14  Landsize       13580 non-null  float64
 15  BuildingArea   13580 non-null  float64
 16  YearBuilt      13580 non-null  int64  
 17  CouncilArea    12211 non-null  object 
 18  Lattit

In [33]:
melb_data.describe().loc[:, ['Distance', 'BuildingArea' , 'Price']]

Distance  BuildingArea         Price
count  13580.000000  13580.000000  1.358000e+04
mean      10.137776    139.633972  1.075684e+06
std        5.868725    392.217403  6.393107e+05
min        0.000000      0.000000  8.500000e+04
25%        6.100000    122.000000  6.500000e+05
50%        9.200000    126.000000  9.030000e+05
75%       13.000000    129.940000  1.330000e+06
max       48.100000  44515.000000  9.000000e+06

In [34]:
melb_data.describe(include=['object'])

Suburb           Address   Type Method SellerG        Date  \
count       13580             13580  13580  13580   13580       13580   
unique        314             13378      3      5     268          58   
top     Reservoir  36 Aberfeldie St      h      S  Nelson  27/05/2017   
freq          359                 3   9449   9022    1565         473   

       CouncilArea             Regionname         Coordinates  
count        12211                  13580               13580  
unique          33                      8               13097  
top       Moreland  Southern Metropolitan  -37.8361, 144.9966  
freq          1163                   4695                  12

In [35]:
melb_data['Regionname'].value_counts()

Southern Metropolitan         4695
Northern Metropolitan         3890
Western Metropolitan          2948
Eastern Metropolitan          1471
South-Eastern Metropolitan     450
Eastern Victoria                53
Northern Victoria               41
Western Victoria                32
Name: Regionname, dtype: int64

In [36]:
melb_data['Regionname'].value_counts(normalize=True)

Southern Metropolitan         0.345729
Northern Metropolitan         0.286451
Western Metropolitan          0.217084
Eastern Metropolitan          0.108321
South-Eastern Metropolitan    0.033137
Eastern Victoria              0.003903
Northern Victoria             0.003019
Western Victoria              0.002356
Name: Regionname, dtype: float64

In [37]:
melb_data['Type'].value_counts(normalize=True)

h    0.695803
u    0.222165
t    0.082032
Name: Type, dtype: float64

Работа с csv файлом. Агрегирующие методы.

In [38]:
print(melb_data['Price'].mean())

1075684.079455081


In [39]:
print(melb_data['Car'].max())
# 10

10


In [40]:
rate = 0.12
income = melb_data['Price'].sum() * rate
print('Total income of real estate agencies:', round(income, 2))
# Total income of real estate agencies: 1752934775.88

Total income of real estate agencies: 1752934775.88


In [41]:
landsize_median = melb_data['Landsize'].median() 
landsize_mean =  melb_data['Landsize'].mean()
print(abs(landsize_median - landsize_mean)/landsize_mean)

0.21205713983546193


In [42]:
landsize_m =  melb_data.mean()
print(landsize_m )

index            6.789500e+03
Rooms            2.937997e+00
Price            1.075684e+06
Distance         1.013778e+01
Postcode         3.105302e+03
Bedroom          2.914728e+00
Bathroom         1.534242e+00
Car              1.611856e+00
Landsize         5.584161e+02
BuildingArea     1.396340e+02
YearBuilt        1.966788e+03
Lattitude       -3.780920e+01
Longtitude       1.449952e+02
Propertycount    7.454417e+03
dtype: float64


/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  """Entry point for launching an IPython kernel.


In [43]:
print(melb_data['Rooms'].mode())

0    3
dtype: int64


In [44]:
print(melb_data['Regionname'].mode())

0    Southern Metropolitan
dtype: object


In [45]:
melb_data['Propertycount'].max()

21650

In [46]:
round(melb_data['Distance'].std(),0)

6.0

In [47]:
med_data = melb_data['BuildingArea'].median()
mean_data = melb_data['BuildingArea'].mean()
raznica = abs(mean_data - med_data)/mean_data
print(round(raznica*100,0))
print(med_data, mean_data)

10.0
126.0 139.63397229013256


In [48]:
melb_data.describe().loc[:, ['Bedroom']]

Bedroom
count  13580.000000
mean       2.914728
std        0.965921
min        0.000000
25%        2.000000
50%        3.000000
75%        3.000000
max       20.000000

In [49]:
melb_data['Bedroom'].value_counts(normalize=True)

3     0.434168
2     0.275184
4     0.191532
1     0.050884
5     0.040943
6     0.004639
0     0.001178
7     0.000736
8     0.000368
9     0.000221
20    0.000074
10    0.000074
Name: Bedroom, dtype: float64

In [50]:
mask = melb_data['Price'] > 2000000
display(mask)

0        False
1        False
2        False
3        False
4        False
         ...  
13575    False
13576    False
13577    False
13578     True
13579    False
Name: Price, Length: 13580, dtype: bool

In [51]:
display(melb_data[mask].head())

index       Suburb               Address  Rooms Type      Price Method  \
80     80  Albert Park  112 Beaconsfield Pde      3    h  2850000.0     PI   
85     85  Albert Park     104 Richardson St      4    h  2300000.0      S   
88     88  Albert Park        29 Faussett St      2    h  2120000.0      S   
92     92  Albert Park           2 Dundas Pl      3    h  2615000.0      S   
93     93  Albert Park          23 Finlay St      5    h  2100000.0      S   

     SellerG        Date  Distance  ...  Car  Landsize  BuildingArea  \
80    Buxton   4/03/2017       3.3  ...    0     211.0         198.0   
85  Marshall   7/05/2016       3.3  ...    1     153.0         180.0   
88      Greg  10/09/2016       3.3  ...    1     199.0         107.0   
92    Cayzer  10/12/2016       3.3  ...    1     177.0         181.0   
93      Greg  10/12/2016       3.3  ...    1     237.0         126.0   

    YearBuilt   CouncilArea  Lattitude  Longtitude             Regionname  \
80       1890  Port Phillip   -37.8481    144.9499  Southern Metropolitan   
85       1880  Port Phillip   -37.8447    144.9523  Southern Metropolitan   
88       1900  Port Phillip   -37.8422    144.9554  Southern Metropolitan   
92       1880  Port Phillip   -37.8415    144.9585  Southern Metropolitan   
93       1970  Port Phillip   -37.8436    144.9557  Southern Metropolitan   

    Propertycount         Coordinates  
80           3280  -37.8481, 144.9499  
85           3280  -37.8447, 144.9523  
88           3280  -37.8422, 144.9554  
92           3280  -37.8415, 144.9585  
93           3280  -37.8436, 144.9557  

[5 rows x 23 columns]

In [52]:
melb_data[melb_data['Price'] > 2000000]

index          Suburb               Address  Rooms Type      Price  \
80        80     Albert Park  112 Beaconsfield Pde      3    h  2850000.0   
85        85     Albert Park     104 Richardson St      4    h  2300000.0   
88        88     Albert Park        29 Faussett St      2    h  2120000.0   
92        92     Albert Park           2 Dundas Pl      3    h  2615000.0   
93        93     Albert Park          23 Finlay St      5    h  2100000.0   
...      ...             ...                   ...    ...  ...        ...   
13521  13521  Port Melbourne          44 Garton St      4    t  2455000.0   
13523  13523         Prahran        69 Greville St      4    h  2668000.0   
13553  13553    Surrey Hills          20 Albert Cr      4    h  2720000.0   
13555  13555    Surrey Hills              3 Oak St      4    h  3100000.0   
13578  13578    Williamstown          96 Verdon St      4    h  2500000.0   

      Method   SellerG        Date  Distance  ...  Car  Landsize  \
80        PI    Buxton   4/03/2017       3.3  ...    0     211.0   
85         S  Marshall   7/05/2016       3.3  ...    1     153.0   
88         S      Greg  10/09/2016       3.3  ...    1     199.0   
92         S    Cayzer  10/12/2016       3.3  ...    1     177.0   
93         S      Greg  10/12/2016       3.3  ...    1     237.0   
...      ...       ...         ...       ...  ...  ...       ...   
13521     SP  Marshall  26/08/2017       3.5  ...    2     123.0   
13523      S    Biggin  26/08/2017       4.6  ...    2     383.0   
13553      S       Kay  26/08/2017      10.2  ...    2    1005.0   
13555     VB  Marshall  26/08/2017      10.2  ...    3     832.0   
13578     PI   Sweeney  26/08/2017       6.8  ...    5     866.0   

       BuildingArea  YearBuilt   CouncilArea  Lattitude  Longtitude  \
80            198.0       1890  Port Phillip  -37.84810   144.94990   
85            180.0       1880  Port Phillip  -37.84470   144.95230   
88            107.0       1900  Port Phillip  -37.84220   144.95540   
92            181.0       1880  Port Phillip  -37.84150   144.95850   
93            126.0       1970  Port Phillip  -37.84360   144.95570   
...             ...        ...           ...        ...         ...   
13521           0.0       2010           NaN  -37.83349   144.94840   
13523         126.0       1970           NaN  -37.84879   144.98882   
13553         126.0       1920           NaN  -37.82421   145.10352   
13555         126.0       1970           NaN  -37.83564   145.10919   
13578         157.0       1920           NaN  -37.85908   144.89299   

                  Regionname  Propertycount           Coordinates  
80     Southern Metropolitan           3280    -37.8481, 144.9499  
85     Southern Metropolitan           3280    -37.8447, 144.9523  
88     Southern Metropolitan           3280    -37.8422, 144.9554  
92     Southern Metropolitan           3280    -37.8415, 144.9585  
93     Southern Metropolitan           3280    -37.8436, 144.9557  
...                      ...            ...                   ...  
13521  Southern Metropolitan           8648   -37.83349, 144.9484  
13523  Southern Metropolitan           7717  -37.84879, 144.98882  
13553  Southern Metropolitan           5457  -37.82421, 145.10352  
13555  Southern Metropolitan           5457  -37.83564, 145.10919  
13578   Western Metropolitan           6380  -37.85908, 144.89299  

[1007 rows x 23 columns]

In [53]:
melb_data[melb_data['Rooms'] == 3].shape[0]

5881

In [54]:
melb_data[((melb_data['Rooms'] == 3) | (melb_data['BuildingArea'] > 100)) & (melb_data['Price'] < 300000)].shape[0]
# 68

68

In [55]:
melb_data[melb_data['Type'] == 't']['Rooms'].max()

5

In [56]:
mean_price = melb_data['Price'].mean()
melb_data[melb_data['Price'] > mean_price]['BuildingArea'].median()

126.0

In [57]:
melb_data['Bathroom'].value_counts()

1    7512
2    4974
3     917
4     106
0      34
5      28
6       5
7       2
8       2
Name: Bathroom, dtype: int64

In [58]:
melb_data.describe().loc[:, ['Bathroom']]

Bathroom
count  13580.000000
mean       1.534242
std        0.691712
min        0.000000
25%        1.000000
50%        1.000000
75%        2.000000
max        8.000000

In [59]:
melb_data[melb_data['Bathroom'] == 0].shape[0]

34

In [60]:
melb_data[(melb_data['SellerG'] == 'Nelson') & (melb_data['Price'] > 3000000)].shape[0]

5

In [61]:
melb_data[melb_data['BuildingArea'] == 0]['Price'].min()

412500.0

In [62]:
round(melb_data[(melb_data['Price'] < 1000000) & ((melb_data['Rooms'] > 5 ) | (melb_data['YearBuilt'] > 2015))]['Price'].mean(),0)

769239.0

In [63]:
melb_data[(melb_data['Type'] == 'h') & (melb_data['Price'] < 3000000)]['Regionname'].value_counts()

Northern Metropolitan         2737
Southern Metropolitan         2520
Western Metropolitan          2286
Eastern Metropolitan          1167
South-Eastern Metropolitan     387
Eastern Victoria                50
Northern Victoria               41
Western Victoria                32
Name: Regionname, dtype: int64

# Закрепление знаний

In [64]:
df_students = pd.read_csv('/content/data/students_performance.csv', sep=',')
display(df_students)

gender race/ethnicity parental level of education         lunch  \
0    female        group B           bachelor's degree      standard   
1    female        group C                some college      standard   
2    female        group B             master's degree      standard   
3      male        group A          associate's degree  free/reduced   
4      male        group C                some college      standard   
..      ...            ...                         ...           ...   
995  female        group E             master's degree      standard   
996    male        group C                 high school  free/reduced   
997  female        group C                 high school  free/reduced   
998  female        group D                some college      standard   
999  female        group D                some college  free/reduced   

    test preparation course  math score  reading score  writing score  
0                      none          72             72             74  
1                 completed          69             90             88  
2                      none          90             95             93  
3                      none          47             57             44  
4                      none          76             78             75  
..                      ...         ...            ...            ...  
995               completed          88             99             95  
996                    none          62             55             55  
997               completed          59             71             65  
998               completed          68             78             77  
999                    none          77             86             86  

[1000 rows x 8 columns]

In [65]:
df_students.shape

(1000, 8)

In [66]:
df_students.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   gender                       1000 non-null   object
 1   race/ethnicity               1000 non-null   object
 2   parental level of education  1000 non-null   object
 3   lunch                        1000 non-null   object
 4   test preparation course      1000 non-null   object
 5   math score                   1000 non-null   int64 
 6   reading score                1000 non-null   int64 
 7   writing score                1000 non-null   int64 
dtypes: int64(3), object(5)
memory usage: 62.6+ KB


In [67]:
df_students.describe()

math score  reading score  writing score
count  1000.00000    1000.000000    1000.000000
mean     66.08900      69.169000      68.054000
std      15.16308      14.600192      15.195657
min       0.00000      17.000000      10.000000
25%      57.00000      59.000000      57.750000
50%      66.00000      70.000000      69.000000
75%      77.00000      79.000000      79.000000
max     100.00000     100.000000     100.000000

In [68]:
df_students['race/ethnicity'].value_counts()

group C    319
group D    262
group B    190
group E    140
group A     89
Name: race/ethnicity, dtype: int64

In [69]:
df_students[df_students['test preparation course'] == 'completed']['writing score'].mean()

74.41899441340782

In [70]:
df_students[df_students['math score'] == 0].shape[0]

1

In [71]:
mean_math_oplata = df_students[df_students['lunch'] == 'standard']['math score'].mean()
mean_math_free = df_students[df_students['lunch'] == 'free/reduced']['math score'].mean()
print(mean_math_oplata, mean_math_free )

70.03410852713178 58.92112676056338


In [72]:
df_students['parental level of education'].value_counts(normalize = True)

some college          0.226
associate's degree    0.222
high school           0.196
some high school      0.179
bachelor's degree     0.118
master's degree       0.059
Name: parental level of education, dtype: float64

In [73]:
zn1 = df_students[df_students['race/ethnicity'] == 'group A']['writing score'].median()
zn2 = df_students[df_students['race/ethnicity'] == 'group C']['writing score'].mean()
print(round(abs(zn1 - zn2),0))

6.0


# Базовые приемы работы с данными в Pandas


In [74]:
melb_df = melb_data.copy()
melb_df.head()

index      Suburb           Address  Rooms Type      Price Method SellerG  \
0      0  Abbotsford      85 Turner St      2    h  1480000.0      S  Biggin   
1      1  Abbotsford   25 Bloomburg St      2    h  1035000.0      S  Biggin   
2      2  Abbotsford      5 Charles St      3    h  1465000.0     SP  Biggin   
3      3  Abbotsford  40 Federation La      3    h   850000.0     PI  Biggin   
4      4  Abbotsford       55a Park St      4    h  1600000.0     VB  Nelson   

        Date  Distance  ...  Car  Landsize  BuildingArea  YearBuilt  \
0  3/12/2016       2.5  ...    1     202.0         126.0       1970   
1  4/02/2016       2.5  ...    0     156.0          79.0       1900   
2  4/03/2017       2.5  ...    0     134.0         150.0       1900   
3  4/03/2017       2.5  ...    1      94.0         126.0       1970   
4  4/06/2016       2.5  ...    2     120.0         142.0       2014   

   CouncilArea  Lattitude  Longtitude             Regionname  Propertycount  \
0        Yarra   -37.7996    144.9984  Northern Metropolitan           4019   
1        Yarra   -37.8079    144.9934  Northern Metropolitan           4019   
2        Yarra   -37.8093    144.9944  Northern Metropolitan           4019   
3        Yarra   -37.7969    144.9969  Northern Metropolitan           4019   
4        Yarra   -37.8072    144.9941  Northern Metropolitan           4019   

          Coordinates  
0  -37.7996, 144.9984  
1  -37.8079, 144.9934  
2  -37.8093, 144.9944  
3  -37.7969, 144.9969  
4  -37.8072, 144.9941  

[5 rows x 23 columns]

In [75]:
melb_df = melb_df.drop(['index', 'Coordinates'], axis=1)
melb_df.head()

Suburb           Address  Rooms Type      Price Method SellerG  \
0  Abbotsford      85 Turner St      2    h  1480000.0      S  Biggin   
1  Abbotsford   25 Bloomburg St      2    h  1035000.0      S  Biggin   
2  Abbotsford      5 Charles St      3    h  1465000.0     SP  Biggin   
3  Abbotsford  40 Federation La      3    h   850000.0     PI  Biggin   
4  Abbotsford       55a Park St      4    h  1600000.0     VB  Nelson   

        Date  Distance  Postcode  ...  Bathroom  Car  Landsize  BuildingArea  \
0  3/12/2016       2.5      3067  ...         1    1     202.0         126.0   
1  4/02/2016       2.5      3067  ...         1    0     156.0          79.0   
2  4/03/2017       2.5      3067  ...         2    0     134.0         150.0   
3  4/03/2017       2.5      3067  ...         2    1      94.0         126.0   
4  4/06/2016       2.5      3067  ...         1    2     120.0         142.0   

   YearBuilt  CouncilArea Lattitude  Longtitude             Regionname  \
0       1970        Yarra  -37.7996    144.9984  Northern Metropolitan   
1       1900        Yarra  -37.8079    144.9934  Northern Metropolitan   
2       1900        Yarra  -37.8093    144.9944  Northern Metropolitan   
3       1970        Yarra  -37.7969    144.9969  Northern Metropolitan   
4       2014        Yarra  -37.8072    144.9941  Northern Metropolitan   

  Propertycount  
0          4019  
1          4019  
2          4019  
3          4019  
4          4019  

[5 rows x 21 columns]

In [76]:
total_rooms = melb_df['Rooms'] + melb_df['Bedroom'] + melb_df['Bathroom']
display(total_rooms)

0         5
1         5
2         8
3         8
4         8
         ..
13575    10
13576     8
13577     8
13578     9
13579     9
Length: 13580, dtype: int64

In [77]:
melb_df['MeanRoomsSquare'] = melb_df['BuildingArea'] / total_rooms
display(melb_df['MeanRoomsSquare'])

0        25.200000
1        15.800000
2        18.750000
3        15.750000
4        17.750000
           ...    
13575    12.600000
13576    16.625000
13577    15.750000
13578    17.444444
13579    12.444444
Name: MeanRoomsSquare, Length: 13580, dtype: float64

In [78]:
diff_area = melb_df['BuildingArea'] - melb_df['Landsize']
sum_area = melb_df['BuildingArea'] + melb_df['Landsize']
melb_df['AreaRatio'] = diff_area/sum_area
display(melb_df['AreaRatio'])

0       -0.231707
1       -0.327660
2        0.056338
3        0.145455
4        0.083969
           ...   
13575   -0.676093
13576   -0.429185
13577   -0.551601
13578   -0.693060
13579   -0.527426
Name: AreaRatio, Length: 13580, dtype: float64

In [79]:
import pandas as pd

def delete_columns(df, col=[]):
    df_columns = df.columns
    for elem in col:
        if elem not in df_columns:
            return None
        else:
            return df.drop(col, axis = 1)

if __name__ == '__main__':
    customer_df = pd.DataFrame({
        'number': [0, 1, 2, 3, 4],
        'cust_id': [128, 1201, 9832, 4392, 7472],
        'cust_age': [13, 21, 19, 21, 60],
        'cust_sale': [0, 0, 0.2, 0.15, 0.3],
        'cust_year_birth': [2008, 2000, 2002, 2000, 1961],
        'cust_order': [1400, 14142, 900, 1240, 8430]
    })
    columns_for_delete= [] #выбранные вами столбцы
    new_df = delete_columns(customer_df, columns_for_delete)
    print(new_df)


None


In [80]:
countries_df = pd.DataFrame({
    'country': ['Англия', 'Канада', 'США', 'Россия', 'Украина', 'Беларусь', 'Казахстан'],
    'population': [56.29, 38.05, 322.28, 146.24, 45.5, 9.5, 17.04],
    'square': [133396, 9984670, 9826630, 17125191, 603628, 207600, 2724902]
})
countries_df['plotnost'] = countries_df['population'] * 1000000 / countries_df['square']
display(countries_df)
display(round(countries_df['plotnost'].mean(),2))

country  population    square    plotnost
0     Англия       56.29    133396  421.976671
1     Канада       38.05   9984670    3.810842
2        США      322.28   9826630   32.796595
3     Россия      146.24  17125191    8.539467
4    Украина       45.50    603628   75.377550
5   Беларусь        9.50    207600   45.761079
6  Казахстан       17.04   2724902    6.253436

84.93

Дата и время

In [81]:
display(melb_df['Date'])

0         3/12/2016
1         4/02/2016
2         4/03/2017
3         4/03/2017
4         4/06/2016
            ...    
13575    26/08/2017
13576    26/08/2017
13577    26/08/2017
13578    26/08/2017
13579    26/08/2017
Name: Date, Length: 13580, dtype: object

In [82]:
melb_df['Date'] = pd.to_datetime(melb_df['Date'], dayfirst=True)
display(melb_df['Date'])

0       2016-12-03
1       2016-02-04
2       2017-03-04
3       2017-03-04
4       2016-06-04
           ...    
13575   2017-08-26
13576   2017-08-26
13577   2017-08-26
13578   2017-08-26
13579   2017-08-26
Name: Date, Length: 13580, dtype: datetime64[ns]

In [83]:
years_sold = melb_df['Date'].dt.year
print(years_sold)
print('Min year sold:', years_sold.min())
print('Max year sold:', years_sold.max())
print('Mode year sold:', years_sold.mode()[0])

0        2016
1        2016
2        2017
3        2017
4        2016
         ... 
13575    2017
13576    2017
13577    2017
13578    2017
13579    2017
Name: Date, Length: 13580, dtype: int64
Min year sold: 2016
Max year sold: 2017
Mode year sold: 2017


In [84]:
melb_df['MonthSale'] = melb_df['Date'].dt.month
melb_df['MonthSale'].value_counts(normalize=True)

5     0.149411
7     0.145950
9     0.135862
6     0.134757
8     0.114138
11    0.082032
4     0.069882
3     0.049926
12    0.044698
10    0.040574
2     0.032622
1     0.000147
Name: MonthSale, dtype: float64

In [85]:
#интервалы
delta_days = melb_df['Date'] - pd.to_datetime('2016-01-01') 
display(delta_days)

0       337 days
1        34 days
2       428 days
3       428 days
4       155 days
          ...   
13575   603 days
13576   603 days
13577   603 days
13578   603 days
13579   603 days
Name: Date, Length: 13580, dtype: timedelta64[ns]

In [86]:
display(delta_days.dt.days)

0        337
1         34
2        428
3        428
4        155
        ... 
13575    603
13576    603
13577    603
13578    603
13579    603
Name: Date, Length: 13580, dtype: int64

In [87]:
melb_df['AgeBuilding'] = melb_df['Date'].dt.year - melb_df['YearBuilt']
display(melb_df['AgeBuilding'])

0         46
1        116
2        117
3         47
4          2
        ... 
13575     36
13576     22
13577     20
13578     97
13579     97
Name: AgeBuilding, Length: 13580, dtype: int64

In [88]:
melb_df = melb_df.drop('YearBuilt', axis=1)

4. Создание и преобразование столбцов с помощью функций

In [89]:
print(melb_df['Address'].nunique())


13378


In [90]:
#несколько строк столбца адрес
print(melb_df['Address'].iloc[177])
print(melb_df['Address'].loc[1812])
print(melb_df['Address'].loc[9001])
# 2/119 Railway St N
# 9/400 Dandenong Rd
# 172 Danks St

2/119 Railway St N
9/400 Dandenong Rd
172 Danks St


In [91]:
# На вход данной функции поступает строка с адресом.
def get_street_type(address):
# Создаём список географических пометок exclude_list.
    exclude_list = ['N', 'S', 'W', 'E']
# Метод split() разбивает строку на слова по пробелу.
# В результате получаем список слов в строке и заносим его в переменную address_list.
    address_list = address.split(' ')
# Обрезаем список, оставляя в нём только последний элемент,
# потенциальный подтип улицы, и заносим в переменную street_type.
    street_type = address_list[-1]
# Делаем проверку на то, что полученный подтип является географической пометкой.
# Для этого проверяем его на наличие в списке exclude_list.
    if street_type in exclude_list:
# Если переменная street_type является географической пометкой,
# переопределяем её на второй элемент с конца списка address_list.
        street_type = address_list[-2]
# Возвращаем переменную street_type, в которой хранится подтип улицы.
    return street_type

In [92]:
text = '2/119 Railway St N'
text.split(' ')

['2/119', 'Railway', 'St', 'N']

In [93]:
street_types = melb_df['Address'].apply(get_street_type)
display(street_types)

0        St
1        St
2        St
3        La
4        St
         ..
13575    Cr
13576    Dr
13577    St
13578    St
13579    St
Name: Address, Length: 13580, dtype: object

In [94]:
print(street_types.nunique())
# 56

56


In [95]:
display(street_types.value_counts())

St           8012
Rd           2825
Ct            612
Dr            447
Av            321
Gr            311
Pde           211
Pl            169
Cr            152
Cl            100
La             67
Bvd            53
Tce            47
Wy             40
Avenue         40
Cct            25
Hwy            24
Parade         15
Boulevard      13
Sq             11
Crescent        9
Cir             7
Strand          7
Esplanade       6
Grove           5
Gdns            4
Grn             4
Fairway         4
Mews            4
Crossway        3
Righi           3
Victoria        2
Ridge           2
Crofts          2
Esp             2
Glade           1
Gra             1
Ave             1
Woodland        1
Outlook         1
Hts             1
Highway         1
Athol           1
Summit          1
Grand           1
Res             1
Nook            1
Eyrie           1
Dell            1
East            1
Loop            1
Grange          1
Terrace         1
Cove            1
Qy              1
Corso     

In [97]:
popular_stypes = street_types.value_counts().nlargest(10).index
print(popular_stypes)
# Index(['St', 'Rd', 'Ct', 'Dr', 'Av', 'Gr', 'Pde', 'Pl', 'Cr', 'Cl'], dtype='object')

Index(['St', 'Rd', 'Ct', 'Dr', 'Av', 'Gr', 'Pde', 'Pl', 'Cr', 'Cl'], dtype='object')


In [98]:
melb_df['StreetType'] = street_types.apply(lambda x: x if x in popular_stypes else 'other')
display(melb_df['StreetType'])

0           St
1           St
2           St
3        other
4           St
         ...  
13575       Cr
13576       Dr
13577       St
13578       St
13579       St
Name: StreetType, Length: 13580, dtype: object

In [99]:
print(melb_df['StreetType'].nunique())
# 11

11


In [100]:
display(melb_df)

Suburb           Address  Rooms Type      Price Method  \
0         Abbotsford      85 Turner St      2    h  1480000.0      S   
1         Abbotsford   25 Bloomburg St      2    h  1035000.0      S   
2         Abbotsford      5 Charles St      3    h  1465000.0     SP   
3         Abbotsford  40 Federation La      3    h   850000.0     PI   
4         Abbotsford       55a Park St      4    h  1600000.0     VB   
...              ...               ...    ...  ...        ...    ...   
13575  Wheelers Hill      12 Strada Cr      4    h  1245000.0      S   
13576   Williamstown     77 Merrett Dr      3    h  1031000.0     SP   
13577   Williamstown       83 Power St      3    h  1170000.0      S   
13578   Williamstown      96 Verdon St      4    h  2500000.0     PI   
13579     Yarraville        6 Agnes St      4    h  1285000.0     SP   

        SellerG       Date  Distance  Postcode  ...  CouncilArea  Lattitude  \
0        Biggin 2016-12-03       2.5      3067  ...        Yarra  -37.79960   
1        Biggin 2016-02-04       2.5      3067  ...        Yarra  -37.80790   
2        Biggin 2017-03-04       2.5      3067  ...        Yarra  -37.80930   
3        Biggin 2017-03-04       2.5      3067  ...        Yarra  -37.79690   
4        Nelson 2016-06-04       2.5      3067  ...        Yarra  -37.80720   
...         ...        ...       ...       ...  ...          ...        ...   
13575     Barry 2017-08-26      16.7      3150  ...          NaN  -37.90562   
13576  Williams 2017-08-26       6.8      3016  ...          NaN  -37.85927   
13577     Raine 2017-08-26       6.8      3016  ...          NaN  -37.85274   
13578   Sweeney 2017-08-26       6.8      3016  ...          NaN  -37.85908   
13579   Village 2017-08-26       6.3      3013  ...          NaN  -37.81188   

       Longtitude                  Regionname  Propertycount MeanRoomsSquare  \
0       144.99840       Northern Metropolitan           4019       25.200000   
1       144.99340       Northern Metropolitan           4019       15.800000   
2       144.99440       Northern Metropolitan           4019       18.750000   
3       144.99690       Northern Metropolitan           4019       15.750000   
4       144.99410       Northern Metropolitan           4019       17.750000   
...           ...                         ...            ...             ...   
13575   145.16761  South-Eastern Metropolitan           7392       12.600000   
13576   144.87904        Western Metropolitan           6380       16.625000   
13577   144.88738        Western Metropolitan           6380       15.750000   
13578   144.89299        Western Metropolitan           6380       17.444444   
13579   144.88449        Western Metropolitan           6543       12.444444   

       AreaRatio  MonthSale AgeBuilding  StreetType  
0      -0.231707         12          46          St  
1      -0.327660          2         116          St  
2       0.056338          3         117          St  
3       0.145455          3          47       other  
4       0.083969          6           2          St  
...          ...        ...         ...         ...  
13575  -0.676093          8          36          Cr  
13576  -0.429185          8          22          Dr  
13577  -0.551601          8          20          St  
13578  -0.693060          8          97          St  
13579  -0.527426          8          97          St  

[13580 rows x 25 columns]

In [101]:
melb_df = melb_df.drop('Address', axis=1)

In [ ]:
#в данном примере, необходимо еще довести до ума моменты замены Avenue на av и другие похожие подтипы

4.2 Ранее, в задании 3.3, мы создали признак WeekdaySale в таблице melb_df — день недели продажи. Из полученных в задании результатов можно сделать вывод, что объекты недвижимости в Мельбурне продаются преимущественно по выходным (суббота и воскресенье).
Напишите функцию get_weekend(weekday), которая принимает на вход элемент столбца WeekdaySale и возвращает 1, если день является выходным, и 0 — в противном случае, и создайте столбец Weekend в таблице melb_df с помощью неё.

Примените эту функцию к столбцу и вычислите среднюю цену объекта недвижимости, проданного в выходные дни. Результат округлите до целых

4.3 Преобразуйте столбец SellerG с наименованиями риелторских компаний в таблице melb_df следующим образом: оставьте в столбце только 49 самых популярных компаний, а остальные обозначьте как 'other'.
Найдите, во сколько раз минимальная цена объектов недвижимости, проданных компанией 'Nelson', больше минимальной цены объектов, проданных компаниями, обозначенными как 'other'. Ответ округлите до десятых.

4.4 Представьте, что вы занимаетесь подготовкой данных о вакансиях с платформы hh.ru. В вашем распоряжении имеется таблица, в которой с помощью парсинга собраны резюме кандидатов. В этой таблице есть текстовый столбец «Опыт работы». Пример такого столбца представлен ниже в виде объекта Series. Структура текста в столбце фиксирована и не может измениться.

Напишите функцию get_experience(arg), аргументом которой является строка столбца с опытом работы. Функция должна возвращать опыт работы в месяцах. Не забудьте привести результат к целому числу. подробнее читать